zebra-state/src/config.rs

```diff
@@ -108,6 +118,73 @@ impl Default for Config {
             cache_dir,
             ephemeral: false,
             debug_stop_at_height: None,
+            delete_old_database: true,
+        }
+    }
+}
+
+/// Check if there are old database folders and delete them from the filesystem.
+///
+/// Iterate over the files and directories in the databases folder and delete if:
+/// - The state directory exists.
+/// - The entry is a directory.
+/// - The directory name has a length of at least 2 characters.
+/// - The directory name has a prefix `v`.
+/// - The directory name without the prefix can be parsed as an unsigned number.
+/// - The parsed number is lower than the hardcoded `DATABASE_FORMAT_VERSION`.
+pub async fn check_and_delete_old_databases(config: Config) {
+    if config.ephemeral || !config.delete_old_database {
+        return;
+    }
+
+    info!("checking old database versions");
+    let cache_dir = config.cache_dir.join("state");
+    if let Some(read_dir) = read_dir(cache_dir.clone()) {
+        for entry in read_dir.flatten() {
+            if let Some(dir_name) = parse_dir_name(entry) {
+                if let Some(version_number) = parse_version_number(dir_name.clone()) {
+                    if version_number < crate::constants::DATABASE_FORMAT_VERSION {
+                        let delete_path = cache_dir.join(dir_name);
+                        if remove_dir_all(delete_path.clone()).is_ok() {
```

Comment:
If we get the `delete_path` wrong, we could delete the user's entire home directory. (Or the entire disk, if Zebra is running as the root user.)

This might also happen if they have a weird cache set up using symlinks.

Here are some things we could do:
- a runtime check that the path is inside the cache directory ([canonicalize](https://docs.rs/tokio/latest/tokio/fs/fn.canonicalize.html), [starts_with](https://doc.rust-lang.org/nightly/std/path/struct.PathBuf.html#method.starts_with))
- an acceptance test that makes sure we delete an old directory, but don't delete files outside the cache directory

What do you think? I can help write these if you want.

Reply:
I dont think i will add more to this PR today so feel free to add this.

Reply:
I wrote the runtime check in 4249029734c3b144184bc05877b9402fa02c28dd.

I didn't have time to write the tests, there are some other things I need to fix today. But I'm happy to help tomorrow.

Did you want to do a simple test?
We could create a fake state directory, there doesn't need to be anything in it.

Reply:
I am not sure of what exactly the test should do. Create a state directory and then run zebra ? Then make sure the directory is not deleted ? Add some old version directories inside and make sure they do get deleted ?

Reply:
> Create a state directory and then run zebra ? Then make sure the directory is not deleted ?

The restart/relaunch tests already do this, so we don't need to add any extra tests for this.

> Add some old version directories inside and make sure they do get deleted ?

This would be good, and also make sure we don't delete files or directories outside the cache directory.

Reply:
How do you know no file in the filesystem was deleted ?

Reply:
Maybe put a `v1` dir anywhere and make sure it was not deleted ?

Reply:
Sounds good!

Maybe put it next to the `state` directory?
(A `v1` directory inside the `state` directory gets deleted. But if it's just one directory lower, it should stay there.)

Reply:
added test at https://github.com/ZcashFoundation/zebra/pull/4586/commits/5a707e1fad1028dc3de6236fb56970f307bb65c4